src/k2/module/k23si/Module.cpp

```diff
@@ -350,6 +578,13 @@ K23SIPartitionModule::handleWrite(dto::K23SIWriteRequest&& request, dto::K23SI_M
                 });
         }
     }
+    
+    // Clean up if req and WI are the same transaction
```

Comment:
This should fix the bug of write same record in a transaction three or more times. But let's move it into above    
 if (versions.size() > 0 && versions[0].status == dto::DataRecord::WriteIntent) { 
Remove previous write in memory if it is written same transaction, otherwise to the rest of check including
if (sitMTR == rec.txnId.mtr) {
            K2DEBUG("Partition: " << _partition << ", post-push winner for key " << request.key);
            // this is a post-PUSH request which won over the siting WI and we still have the WI in cache
            _queueWICleanup(std::move(rec));
            versions.pop_front();
        }
        else if (rec.txnId.mtr != rqmtr) {

Also, I suggest that we can optimize this function related with push. First remove the parameter sitMtr from the API, it is kind of confusing. Then inside request (or mtr cache when we have one), we can keep a list of txnId(key) this request transaction already wins through push, so that we can check against this list first when we need to push, instead of using sitMtr for it. This way, assuming the list was maintained on the request transaction, we can avoid push against other writeintent of the transactions we already pushed and won before. This optimization can be done later and I will sync with Ivan about this as well.